vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestExecutor.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.http;
+
+import java.io.IOException;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import org.apache.http.ConnectionClosedException;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.vividus.http.client.HttpResponse;
+import org.vividus.http.client.IHttpClient;
+import org.vividus.http.exception.HttpRequestBuildException;
+import org.vividus.softassert.ISoftAssert;
+
+public class HttpRequestExecutor
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpRequestExecutor.class);
+
+    private final IHttpClient httpClient;
+    private final HttpTestContext httpTestContext;
+    private final ISoftAssert softAssert;
+
+    public HttpRequestExecutor(IHttpClient httpClient, HttpTestContext httpTestContext, ISoftAssert softAssert)
+    {
+        this.httpClient = httpClient;
+        this.httpTestContext = httpTestContext;
+        this.softAssert = softAssert;
+    }
+
+    public void executeHttpCall(HttpMethod httpMethod, String endpoint, String relativeURL)
+    {
+        executeHttpCall(Supplier::get, httpMethod, endpoint, relativeURL);
+    }
+
+    public void executeHttpCall(Consumer<Supplier<HttpResponse>> consumer,
+            HttpMethod httpMethod, String endpoint, String relativeURL)
+    {
+        try
+        {
+            consumer.accept(() -> executeHttpCallSafely(httpMethod, endpoint, relativeURL));
+        }
+        finally
+        {
+            httpTestContext.releaseRequestData();
+        }
+    }
+
+    private HttpResponse executeHttpCallSafely(HttpMethod httpMethod, String endpoint, String relativeURL)
+    {
+        try
+        {
+            return execute(httpMethod, endpoint, relativeURL);
+        }
+        catch (HttpRequestBuildException | ConnectionClosedException e)
+        {
+            softAssert.recordFailedAssertion(e);
+            return null;
+        }
+        catch (IOException io)
+        {
+            throw new HttpRequestException("An I/O exception occurred", io);
```

Comment:
there is no much profit to wrap this exception, moreover examples below shows that developers must handle `RuntimeException`-s, it's rather harder to track unchecked exceptions than checked ones

Reply:
https://github.com/vividus-framework/vividus/pull/59#discussion_r329116182

Throwing IOException in this method will affect downstream method 
`HttpRequestExecutor.executeHttpCall(Consumer<Runnable> consumer, HttpMethod httpMethod, String endpoint,
            String relativeURL)`
which require exception handling inside consumer. I thought it is better to wrap it here. Please, advise more elegant code.

Reply:
Ivan't comment was about method naming `...Safely` - method can be safe if it throws exception.


> which require exception handling inside consumer.

do not perform exception handling, just pass it to calling methods

Reply:
> Ivan't comment was about method naming ...Safely - method can be safe if it throws exception.
Ok, now clear - this was initial method in ApiSteps. 

> do not perform exception handling, just pass it to calling methods
Consumer require handling anyway here: 
https://github.com/vividus-framework/vividus/pull/59/files#diff-89f724ebef795aa0c8c7e12db89755f8R60

Reply:
```java
@FunctionalInterface
public interface CheckedIOConsumer<T>
{
    void accept(T t) throws IOException;
}
```

Reply:
Requires CheckedIORunnable as well.